### DISCUSSION TITLE
Valid mcap file with nested arrays does not load

### DISCUSSION THREAD

**Discussion by defunctzombie** [foxglove/community#583] [Q&A]
**Description**

I have an mcap file with a valid json schema and message which fails to open in studio but does work with the mcap CLI tool.

<img width="725" alt="image" src="https://user-images.githubusercontent.com/84792/209162936-7483c26d-3c97-4916-a00c-80fc4e8e107b.png">

[simple_file.mcap.zip](https://github.com/foxglove/studio/files/10287893/simple_file.mcap.zip)

```
library: python mcap 0.0.16
profile:
messages: 1
duration: 0s
start: 1.000
end: 1.000
compression:
	zstd: [1/1 chunks] [00 MB/00 MB (38.13%)]
channels:
  	(1) matrix_data  1 msgs (+Inf Hz)   : matrix_schema [jsonschema]
attachments: 0
metadata: 0
```

```
(base) ➜  Downloads mcap cat simple_file.mcap --json
{"topic":"matrix_data","sequence":0,"log_time":1.000000000,"publish_time":1.000000000,"data":{"matrix":[[1,0,0]]}}
```

```
  1   matrix_schema  jsonschema  {"$schema": "http://json-schema.org/draft-04/schema#", "type": "object", "properties": {"matrix": {"type": "array", "minItems": 1, "maxItems": 1, "items": {"type": "array", "items": {"type": "integer"}, "minItems": 3, "maxItems": 3}}}}
```

**Expected Behavior**
The mcap file loads without errors and I can see the matrix_data in a raw message panel.

**Comment by jtbandes** [comment]
Studio does not and has never supported nested arrays. I think this is a feature request. Message path syntax may have to be changed if we want to support them.

**Reply by sachitv** [reply]
Related question:
Is foxglove studio by design expected to be able to load any validly encoded (protobuf, json, flatbuffer etc) mcap file or is it restricted to only mcap files matching foxglove studio schemas?

I’ve been under the impression that it’s the former and that I can write message converter extensions to transform / convert my message to foxglove schemas. Have I been making an incorrect assumption?

It would be great if it was possible to load some types but not necessarily be able to visualise them given the possibility of doing transformations on the data. Perhaps it would be a good strategy allow loading of unsupported types without visualisation support as opposed to failing to load them from the mcap file?

**Reply by defunctzombie** [reply]
In Studio we aim to support ros, protobuf, flatbuffer, and json serializations. However this comes with some caveats. Studio is currently designed around binary serialization formats and most of the ones we support are not as flexible as the types of messages you can create with JSON. Nested arrays is one good example. I did some quick searching and it seems that in protobuf the approach is to use one flat array - there is no array nesting. This is also true in ROS and I believe flatbuffer. As a result, we don't support this for JSON messages either.

In the case of nested arrays there are some internal representations we'd need to relax or change. One thing you could try is to use an empty json schema (or one that is simply an object with any fields ... I think `{ type: object, additionalProperties: true }` should suffice for that. Studio won't do any schema validation, you won't get autocomplete in message paths, but I believe studio will de-serialize and display your messages in various panels like Raw Message and Plot and you should be able to write message converters as well. The main issue you are running into is not one of deserialization but of schema validation so if you supply a different schema Studio might load the data.

**Reply by sachitv** [reply]
Thanks for that info. I was able to view my messages in the raw panel using a simple object type in the schema.

**Comment by defunctzombie** [comment]
@jtbandes wow yea you are right - I had in my mind that things like the camera info matrix were nested arrays but I see those are single arrays and code is meant to "interpret" them a certain way.

Removing the bug label. Indeed this is a feature request.
